server/chat-commands/avatars.tsx

```diff
@@ -0,0 +1,869 @@
+			throw new Chat.ErrorMessage(`Avatar "${avatar}" is an official avatar that all users already have access to.`);
+		}
+		return avatar;
+	}
+	img(avatar: AvatarID, noAlt?: boolean) {
+		const src = Avatars.src(avatar);
+		if (!src) return <strong><code>{avatar}</code></strong>;
+		return <img
+			src={src} alt={noAlt ? '' : avatar} width="80" height="80" class="pixelated" style={{verticalAlign: 'middle'}}
+		/>;
+	}
+	getDefault(userid: ID) {
+		const entry = customAvatars[userid];
+		if (!entry) return null;
+		const DECEMBER = 11; // famous JavaScript issue
+		if (new Date().getMonth() === DECEMBER && entry.allowed.some(avatar => avatar?.endsWith('xmas'))) {
+			return entry.allowed.find(avatar => avatar?.endsWith('xmas'));
+		}
+		return entry.default === undefined ? entry.allowed[0] : entry.default;
+	}
+	/** does not include validation */
+	setDefault(userid: ID, avatar: AvatarID | null) {
+		if (avatar === this.getDefault(userid)) return;
+
+		const entry = (customAvatars[userid] ??= {allowed: [null]});
+		if (avatar === entry.allowed[0]) {
+			delete entry.default;
+		} else {
+			entry.default = avatar;
+		}
+		saveCustomAvatars();
+	}
+	addAllowed(userid: ID, avatar: AvatarID | null) {
+		const entry = (customAvatars[userid] ??= {allowed: [null]});
+		if (entry.allowed.includes(avatar)) return false;
+
+		entry.allowed.push(avatar);
+		entry.notNotified = true;
+		this.tryNotify(Users.get(userid));
+		return true;
+	}
+	removeAllowed(userid: ID, avatar: AvatarID | null) {
+		const entry = customAvatars[userid];
+		if (!entry?.allowed.includes(avatar)) return false;
+
+		if (entry.allowed[0] === avatar) {
+			entry.allowed[0] = null;
+		} else {
+			entry.allowed = entry.allowed.filter(a => a !== avatar) as any;
+		}
+		if (!entry.allowed.some(Boolean)) delete customAvatars[userid];
+		return true;
+	}
+	addPersonal(userid: ID, avatar: AvatarID | null) {
+		const entry = (customAvatars[userid] ??= {allowed: [null]});
+		if (entry.allowed.includes(avatar)) return false;
+
+		entry.timeReceived ||= Date.now();
+		entry.timeUpdated = Date.now();
+		if (!entry.allowed[0]) {
+			entry.allowed[0] = avatar;
+		} else {
+			entry.allowed.unshift(avatar);
+		}
+		delete entry.default;
+		entry.notNotified = true;
+		this.tryNotify(Users.get(userid));
+		return true;
+	}
+	handleLogin(user: User) {
+		const avatar = this.getDefault(user.id);
+		if (avatar) user.avatar = avatar;
+		this.tryNotify(user);
+	}
+	tryNotify(user?: User | null) {
+		if (!user) return;
+
+		const entry = customAvatars[user.id];
+		if (entry?.notNotified) {
+			user.send(
+				`|pm|&|${user.getIdentity()}|/raw ` +
+				Chat.html`${<>
+					<p>
+						You have a new custom avatar!
+					</p>
+					<p>
+						{entry.allowed.map(avatar => avatar && [Avatars.img(avatar), ' '])}
+					</p>
+					Use <button class="button" name="send" value="/avatars"><code>/avatars</code></button> for usage instructions.
+				</>}`
+			);
+			delete entry.notNotified;
+			saveCustomAvatars();
```

Comment:
I find it a little confusing that the avatars are saved in a function called `tryNotify`.

Reply:
`saveCustomAvatars` saves the entire avatar table `customAvatars`, which tracks whether or not the user has been notified of an avatar. Is there a rename you'd suggest?

Reply:
`tryNotifyAndSave`?

Reply:
I don't think `andSave` is important enough to go in the function name? Why is that important to you?